crates/red_knot_python_semantic/src/semantic_index/symbol.rs

```diff
@@ -3,35 +3,39 @@ use std::ops::Range;
 
 use bitflags::bitflags;
 use hashbrown::hash_map::RawEntryMut;
-use rustc_hash::FxHasher;
-use smallvec::SmallVec;
-
+use ruff_db::parsed::ParsedModule;
 use ruff_db::vfs::VfsFile;
 use ruff_index::{newtype_index, IndexVec};
 use ruff_python_ast::name::Name;
+use ruff_python_ast::{self as ast};
+use rustc_hash::FxHasher;
 
+use crate::ast_node_ref::AstNodeRef;
+use crate::node_key::NodeKey;
 use crate::semantic_index::definition::Definition;
-use crate::semantic_index::{root_scope, semantic_index, symbol_table, NodeWithScopeId, SymbolMap};
+use crate::semantic_index::{root_scope, semantic_index, symbol_table, SymbolMap};
 use crate::Db;
 
 #[derive(Eq, PartialEq, Debug)]
-pub struct Symbol {
+pub struct Symbol<'db> {
     name: Name,
     flags: SymbolFlags,
     /// The nodes that define this symbol, in source order.
-    definitions: SmallVec<[Definition; 4]>,
+    ///
+    /// TODO: Use smallvec here, but it creates the same lifetime issues as in [QualifiedName](https://github.com/astral-sh/ruff/blob/5109b50bb3847738eeb209352cf26bda392adf62/crates/ruff_python_ast/src/name.rs#L562-L569)
+    definitions: Vec<Definition<'db>>,
```

Comment:
FWIW, ultimately with CFG I'm not sure we'll even need this, because what will matter are reachable definitions from a given use, not all definitions of a symbol. This might be relevant because it would remove the db lifetime from the symbol table again.

Reply:
Won't this be necessary when inferring the public type of a symbol where we unify the type of all definitions?

Reply:
I think public type should be "reachable from end of scope" not "all definitions" -- but either way, if we associate uses with a list of definitions, "public type" is just one more "use" -- we don't have to embed all definitions directly in the symbol itself.